src/map/script.c

```diff
@@ -12572,6 +12572,162 @@ static BUILDIN(cloakoffnpc)
 	return true;
 }
 
+
+/*
+ * Create a duplicate of source NPC.
+ * npc_duplicate("<source_npc_name>", "<new_npc_name>", "<new_npc_hidden_name>", "<mapname>", <map_x>, <map_y>, <dir>{, <sprite_id>{, <map_xs>, <map_ys>}});
+ * Return 1 on success, 0 if failed.
+ */
+static BUILDIN(npc_duplicate)
+{
+	int txs = -1, tys = -1;
+
+	if (script_hasdata(st, 10))
+		txs = script_getnum(st, 10);
+	if (script_hasdata(st, 11))
+		tys = script_getnum(st, 11);
+
+	if (txs < -1)
+		txs = -1;
+	if (tys < -1)
+		tys = -1;
+
+	if (txs == -1 && tys != -1)
+		txs = 0;
+	if (txs != -1 && tys == -1)
+		tys = 0;
+
+	const char *dup_name = script_getstr(st, 3);
+	const char *dup_hidden_name = script_getstr(st, 4);
+
+	char targetname[NAME_LENGTH] = "";
+
+	if (strlen(dup_name) > NAME_LENGTH) {
+		ShowError("buildin_npc_duplicate: NPC Duplicated name '%s' is too long (max %d chars).\n", dup_name, NAME_LENGTH);
+		script_pushint(st, 0);
+		return false;
+	}
+
+	strcat(targetname, dup_name);
+
+	if (strlen(dup_hidden_name) != 0) {
+		strncat(targetname, "#", 1);
+		strncat(targetname, dup_hidden_name, strlen(dup_hidden_name));
```

Comment:
This still can overflow since you did not check the total length of `dup_name` + `dup_hidden_name` + 1 (`'#'`).
Do some thing like this before filling `targetname`:
size_t len_dup = strlen(dup_name);
size_t len_dup_hid = strlen(dup_hidden_name);
```
if (len_dup + len_dup_hid + 1 > NAME_LENGTH - 1) {
	// Target name too long... abort duplicating.
}
```
And then use `safesnprintf(targetname, sizeof(targetname), "%s#%s", dup_name, dup_hidden_name);` to fill `targetname` instead of string concatenation.

And add a check if the target name already is in use.
```
if (npc->name2id(targetname) != NULL) {
	// Name in use... abort duplicating or generate new one.
}
```

Reply:
added the length check, the `npc->name2id(...)` already have it.